config_converter/config_converter.py

```diff
@@ -0,0 +1,213 @@
+        specific = {}
+        for p in d.params:
+            if p.name not in outer_map:
+                if p.name in self.unsupported_fields:
+                    print(f'parameter {p.name} cant be converted')
+                elif p.name in special_params:
+                    self._map_special_fields(specific, p, plugin)
+                elif p.name not in main_map:
+                    print(f'parameter {p.name} is unknown')
+                else:
+                    specific[main_map[p.name]] = p.value
+        for nd in d.directives:
+            if nd.name not in supported_dirs:
+                print(f'directive {nd.name} is not supported')
+            else:
+                for np in nd.params:
+                    self._map_special_fields(specific, np, plugin)
+        cur[f'{cur["type"]}_{d.name}_config'] = specific
+        return cur
+
+    def _map_special_fields(self, specific: dict, p: config_pb2.Param,
+                            plugin: str) -> None:
+        """Cases on p with all special_map dict, calls a map function."""
+        if p.name in self.special_map[plugin]:
+            if plugin == 'in_tail':
+                self.parser_dir(specific, p)
+
+    def _create_mapping_info(self) -> None:
+        """Initializes all the mapping rules."""
+        # each sub dict contains fields that stay in the current level only
+        self.main_map = {
+            'in_tail': {
+                'exclude_path': 'exclude_path',
+                'path': 'path',
+                'path_key': 'path_field_name',
+                'pos_file': 'checkpoint_file',
+                'refresh_interval': 'refresh_interval',
+                'rotate_wait': 'rotate_wait'
+            }
+        }
+        # each sub list contains fields that need to be treated differently
+        self.special_map = {
+            'in_tail': [
+                'format', 'format_firstline', '@type',
+                'multiline_flush_interval', 'expression'
+            ] + [f'format{i}' for i in range(1, 21)]
+        }
+        # dictionary of dictionaries - one per main plugin type (source, match)
+        # each dict contains fields that go in the upper level
+        self.outer_map = {'source': {'tag': 'name', '@type': 'type'}}
+        # each sub list contains known and allowed sub directives
+        self.supported_dirs = {'in_tail': ['parse']}
+        self.unsupported_fields = [
+            'emit_unmatched_lines', 'enable_stat_watcher',
+            'enable_watch_timer', 'encoding', 'from_encoding',
+            'ignore_repeated_permission_error', 'limit_recently_modified',
+            'open_on_every_update', 'path_timezone',
+            'pos_file_compaction_interval', 'read_from_head',
+            'read_lines_limit', 'skip_refresh_on_startup'
+        ]
+        self.supported_plugins = ['in_tail']
+        self.unsupported_plugins = ['in_forward', 'in_syslog']
+
+    def parser_dir(self, specific: dict, p: config_pb2.Param) -> None:
```

Comment:
Is `parser_dir` used elsewhere? If not, can we follow the style as private method? I think this is still internal function as it is called by `_map_special_fields` only, but maybe I am wrong.

Reply:
You're right - it is only called from `_map_special_fields` but if I make this private too, then all the methods in the class become private. I'm not sure if that is good style? I might be wrong though.

Reply:
See my another added comment on `write_to_yaml`.